conftest.py

```diff
@@ -0,0 +1,58 @@
+import pytest
+from selenium import webdriver
+
+from api.auth import AuthAPI
+from const import Constants as C
+from data import Data as D
+import tools
+
+from pages.login_page import LoginPage
+from pages.main_page import MainPage
+
+
+@pytest.fixture(scope="session")
+def data():
+    data = D()
+    yield data
+    del data
+
+
+@pytest.fixture(scope="session")
+def test_user():
+    with AuthAPI(C.APP_URL, **C.TEST_USER) as test_user:
+        test_user.register()
+        yield test_user
+
+
+@pytest.fixture(params=["Firefox", "Chrome"], scope="function")
+# @pytest.fixture(params=["Firefox"], scope="function")
+# @pytest.fixture(params=["Chrome"], scope="function")
+def driver(request, test_user):
+    _Driver = getattr(webdriver, request.param)
+    _Options = getattr(webdriver, f"{request.param}Options")
+    options = _Options()
+    options.add_argument("--headless")
+    options.add_argument("--window-size=1280,1280")
+    driver = _Driver(options=options)
+
+    driver.get(C.APP_URL)
+    yield driver
+    driver.quit()
+
+
+@pytest.fixture(scope="function")
+def authorized(driver, test_user):
+    login_page = LoginPage(driver)
+    login_page.open_page()
+    if login_page.is_loaded():
+        login_page.login(email=test_user.email, password=test_user.password)
+        yield driver
+    else:
+        raise RuntimeError("Login failed")
+
+@pytest.fixture(scope="function")
+def user_with_orders(authorized, test_user, data):
+    main_page = MainPage(authorized)
+    tools.make_order(main_page, data)
+    yield authorized
```

Comment:
Можно лучше: тут аналогично с заданием 2, хватило бы return

Reply:
Я немного вникал в этот вопрос. И есть нюансы. Возможно, здесь можно и `return` (Хотя в одной статье про `return` и `yield` в контексте генераторов писалось, что `yield` есть ряд существенных преимуществ). Но во втором дипломном проекте и здесь, где через АПИ создаётся пользователь, `yield` вызывается внутри `with` и в таком ключе есть рекомендация использовать именно `yield`.

[вот тут](https://stackoverflow.com/questions/77318902/why-using-yield-instead-of-return-in-pytest-fixtures) есть такая цитата
>That's because you are in a context manager (`with` clause), and if you use `return`, the control is not returned to the fixture and the cleanup of the context manager is not called (e.g. the patching is not reverted). This can have any unwanted side effects and wil likely fail other tests. So, as @FrankYellin wrote: just use always `yield` in a fixture, so you won't run into this kind of problem.